File: sdklibrary/src/main/java/com/loner/android/sdk/webservice/network/networking/RetrofitClientInstance.java
package com.loner.android.sdk.webservice.network.networking;

import java.util.concurrent.TimeUnit;

import okhttp3.CertificatePinner;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitClientInstance implements RequestConfig {

    private static Retrofit retrofit;

    public static Retrofit getRetrofitInstance() {
        if (retrofit == null) {
            HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
            interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

            CertificatePinner certificatePinner = new CertificatePinner.Builder()
                    .add(HOST_NAME, "sha256/" + SHA_256)
                    .build();

            OkHttpClient client = new OkHttpClient.Builder().addInterceptor(interceptor)
                    .certificatePinner(certificatePinner)
                    .connectTimeout(TIMEOUT, TimeUnit.MILLISECONDS).build();

            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(client)
                    .build();
        }
        return retrofit;
    }
}

File: sdklibrary/src/main/java/com/loner/android/sdk/webservice/network/networking/RetrofitClientInstance.kt
package com.loner.android.sdk.webservice.network.networking

import okhttp3.CertificatePinner
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

class RetrofitClientInstance : RequestConfig {
    companion object {

        private var retrofit: Retrofit? = null

        val retrofitInstance: Retrofit?
            get() {
                if (retrofit == null) {
                    val interceptor = HttpLoggingInterceptor()
                    interceptor.level = HttpLoggingInterceptor.Level.BODY

                    val certificatePinner = CertificatePinner.Builder()
                            .add(RequestConfig.Companion.HOST_NAME, "sha256/" + RequestConfig.Companion.SHA_256)
                            .build()

                    val client = OkHttpClient.Builder().addInterceptor(interceptor)
                            .certificatePinner(certificatePinner)
                            .connectTimeout(RequestConfig.Companion.TIMEOUT.toLong(), TimeUnit.MILLISECONDS).build()

                    retrofit = Retrofit.Builder()
                            .baseUrl(RequestConfig.Companion.BASE_URL)
                            .addConverterFactory(GsonConverterFactory.create())
                            .client(client)
                            .build()
                }
                return retrofit
            }
    }
}
